pandas/io/parquet.py

```diff
@@ -115,17 +122,21 @@ def write(
                 **kwargs,
             )
         else:
-            self.api.parquet.write_table(
-                table, file_obj_or_path, compression=compression, **kwargs
-            )
+            self.api.parquet.write_table(table, path, compression=compression, **kwargs)
```

Comment:
It might have been discussed before in this PR, but why this change from `file_obj_or_path` to `path` ? Because with this change, `file_obj_or_path` is basically never used?

Reply:
You know what, this can be made simpler...

The reason is, pyarrow can take a filesystem+path and do the right thing, and for the partitioning case, this is the only way. I think we can deal without opening the file in the pandas code at all. I'll push that suggestion, and add comments for what the blocks do to be clearer.

Reply:
The new code above certainly looks fine, I am only wondering if there were things that `get_filepath_or_buffer` was doing before we might now be missing.

It gets a buffer for urls, but I don't think this actually works for *writing*. 
In addition, it also seems to do a `os.path.expanduser` on strings. I suppose this is not tested (and not sure many people would rely on that), but that might lead to a regression.

Reply:
You want to add `expanduser` to the function? I wouldn't bother going through `get_filepath_or_buffer` just for this, and as you say, it isn't tested or documented anywhere. In fact, test may be a little tricky for the various OSs, I'm not sure you can guarantee that `"~/afile"`-> `"${HOME}/afile"`.

Reply:
Confirmed that it works on master

```python
In [2]: df = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})

In [3]: df.to_parquet("~/test-4.parquet")
```

and raises on this branch.

```python
In [3]: df.to_parquet("~/test-5.parquet")
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-3-c7dcbbaae894> in <module>
----> 1 df.to_parquet("~/test-5.parquet")

~/sandbox/pandas/pandas/util/_decorators.py in wrapper(*args, **kwargs)
    197                 else:
    198                     kwargs[new_arg_name] = new_arg_value
--> 199             return func(*args, **kwargs)
    200
    201         return cast(F, wrapper)

~/sandbox/pandas/pandas/core/frame.py in to_parquet(self, path, engine, compression, index, partition_cols, **kwargs)
   2323             index=index,
   2324             partition_cols=partition_cols,
-> 2325             **kwargs,
   2326         )
   2327

~/sandbox/pandas/pandas/io/parquet.py in to_parquet(df, path, engine, compression, index, partition_cols, **kwargs)
    266         index=index,
    267         partition_cols=partition_cols,
--> 268         **kwargs,
    269     )
    270

~/sandbox/pandas/pandas/io/parquet.py in write(self, df, path, compression, index, partition_cols, **kwargs)
    118         else:
    119             # write to single output file
--> 120             self.api.parquet.write_table(table, path, compression=compression, **kwargs)
    121
    122     def read(self, path, columns=None, **kwargs):

~/Envs/pandas-dev/lib/python3.7/site-packages/pyarrow/parquet.py in write_table(table, where, row_group_size, version, use_dictionary, compression, write_statistics, use_deprecated_int96_timestamps, coerce_timestamps, allow_truncated_timestamps, data_page_size, flavor, filesystem, compression_level, **kwargs)
   1341                 use_deprecated_int96_timestamps=use_int96,
   1342                 compression_level=compression_level,
-> 1343                 **kwargs) as writer:
   1344             writer.write_table(table, row_group_size=row_group_size)
   1345     except Exception:

~/Envs/pandas-dev/lib/python3.7/site-packages/pyarrow/parquet.py in __init__(self, where, schema, filesystem, flavor, version, use_dictionary, compression, write_statistics, use_deprecated_int96_timestamps, compression_level, **options)
    434         filesystem, path = resolve_filesystem_and_path(where, filesystem)
    435         if filesystem is not None:
--> 436             sink = self.file_handle = filesystem.open(path, 'wb')
    437         else:
    438             sink = where

~/Envs/pandas-dev/lib/python3.7/site-packages/pyarrow/filesystem.py in open(self, path, mode)
    242         """
    243         path = _stringify_path(path)
--> 244         return open(path, mode=mode)
    245
    246     @property

FileNotFoundError: [Errno 2] No such file or directory: '~/test-5.parquet'
```

pandas consistently expands `~` to the homedir, so we'll need to ensure we don't regress on this.

Reply:
but how do we test this reliably?

Reply:
We should have similar tests for csv.

Reply:
We test the common utility: https://github.com/pandas-dev/pandas/blob/80ba4c4294a1ce1d95aeeb8e5de86d33ebb6edf4/pandas/tests/io/test_common.py#L55-L68

and then probaby suppose that all IO methods use those common utilities.

Reply:
A little weak...
I'll put in something. By the way, I expect that expanding the home dir didn't work before either when partitioning.